### PR TITLE
improved filepackaging

### PR DESCRIPTION
**Pull request by DerThorsten**
This allows having a config file to specify what is `packaged` for which package.

patters can be of following types:
 * regex
 * file-extension
 * contains a substring
 
one can have a config as yaml
```yaml
packages:
  python-dateutil: 
    include_patterns:
      - type : regex
        pattern: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)|(.*dateutil-zoneinfo\.tar\.gz$)'
      - type: contains_any : 
        contains_any ['fubar','fu']
    exclude_patterns:
      - type: contains_any : 
        contains_any ['/debug/','/test/']

default: 
  include_patterns:
    - type : regex
      pattern: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)'
```

and we have a sensible default directly in the code

### CODE REVIEW

**Comment by wolfv** [comment]
I really don't like complicated regex.

It seems that most of the examples could be expressed as straight-forward globs?

```
default: 
  include_patterns:
    - **/*.so
    - **/*.py
  exclude_patterns:
    - **/tests/*
```

**Comment by DerThorsten** [comment]
I think we should allow support for regex since with them once can anything.
@wolfv  I will add support for `fnmatch` and make the defaults based on them

**Comment by wolfv** [comment]
cool! 
Ok, let's keep regex. We can also identify a regex if the string starts with `^` and ends with `$`, then we could simplify the YAML.

**Comment by DerThorsten** [comment]
@wolfv  I removed all the filters except the regex one and added the "glob"/"unix-wildcard-style" one.
I prefer to be explicit, so instead of relying on `^` to detect regex, we use the key `regex` for a regex and `pattern` for the UNIX style wildcard patterns.

The yaml now looks like:
``` yaml
packages:
  python-dateutil: 
    include_patterns:
      - regex: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)|(.*dateutil-zoneinfo\.tar\.gz$)'
      - patter: "*.so" 
      - patter: "**/include_me/*"
    exclude_patterns:
      - patter: "*.a" 
      - patter: "**/exclud_me/*"
default: 
  include_patterns:
    - regex: '^(?!.*\/tests\/).*(.*.\.py$)|(.*.\.so$)'
```

**Comment by DerThorsten** [comment]
this PR is also needed to fix https://github.com/jupyterlite/xeus-python-kernel/issues/55#issuecomment-1195108548

File: empack/file_packager.py [hunk 1]
```diff
@@ -224,37 +159,55 @@ def emscripten_file_packager(
         )
 
 
-def pack_python_core(env_prefix: Path, outname, version, export_name, download_emsdk):
-    ensure_python(env_prefix=env_prefix, version=version)
+def pack_environment(
+    env_prefix: Path, outname, export_name, pack_outdir=None, download_emsdk=None
+):
+    # name of the env
+    print("env_prefix", env_prefix)
```
Comment (martinRenou) [inline empack/file_packager.py:166]:
```suggestion
```

File: empack/file_packager.py [hunk 2]
```diff
@@ -224,37 +159,55 @@ def emscripten_file_packager(
...
+    env_name = PurePath(env_prefix).parts[-1]
+    env_root = os.path.join(*PurePath(env_prefix).parts[:-1])
 
-    lib_dir = env_prefix / "lib"
-    python_lib_dir = lib_dir / f"python{version}"
+    print(f"{env_name=} {env_root=}")
```
Comment (martinRenou) [inline empack/file_packager.py:170]:
```suggestion
```

File: empack/file_packager.py [hunk 3]
```diff
@@ -224,37 +159,55 @@ def emscripten_file_packager(
...
-        shutil.copy(os.path.join(temp_dir_str, f"{outname}.js"), os.getcwd())
+
+def pack_python_core(env_prefix: Path, outname, version, export_name, download_emsdk):
+    warnings.warn(
+        "pack_python_core is deprecated, use `pack_environment`",
+        DeprecationWarning,
+    )
+    print(
+        "pack_python_core is deprecated, use `pack_environment`",
+    )
```
Comment (martinRenou) [inline empack/file_packager.py:204]:
```suggestion
```

**Review by martinRenou — COMMENTED** [review]
Thanks! There are some leftover prints

**Comment by DerThorsten** [comment]
> Thanks! There are some leftover prints

thanks for catching

**Comment by wolfv** [comment]
LGTM. I would squash the commits a bit more, but that's all.

**Comment by DerThorsten** [comment]
we can squash everything, I dont mind

**Pull request merged**